test/integration/test_celery_user_rate_limit.py

```diff
@@ -0,0 +1,169 @@
+import datetime
+import tempfile
+from functools import lru_cache
+from typing import (
+    Dict,
+    Iterable,
+    List,
+)
+
+from celery.result import AsyncResult
+from sqlalchemy import text
+
+from galaxy.celery import galaxy_task
+from galaxy.model.database_utils import sqlalchemy_engine
+from galaxy.util import ExecutionTimer
+from galaxy_test.driver.driver_util import init_database
+from galaxy_test.driver.integration_util import IntegrationTestCase
+
+
+@galaxy_task(bind=True)
+def mock_user_id_task(self, task_user_id: int):
+    return task_user_id
+
+
+@lru_cache()
+def sqlite_url():
+    path = tempfile.NamedTemporaryFile().name
+    dburl = f"sqlite:///{path}"
+    init_database(dburl)
+    return dburl
+
+
+@lru_cache()
+def setup_users(dburl: str, num_users: int = 2):
+    """
+    Setup test users in galaxy_user table with user id's starting from 2 because
+    we assume there always exists a user with id = 1.
+    This is because the new celery_user_rate_limit table has
+    a user_id with a foreign key pointing to galaxy_user table.
+    """
+    expected_user_ids = [i for i in range(2, num_users + 1)]
+    with sqlalchemy_engine(dburl) as engine:
+        with engine.begin() as conn:
+            found_user_ids = conn.scalars(
+                text("select id from galaxy_user where id between 1 and :high"), {"high": num_users}
+            ).all()
+            if len(expected_user_ids) > len(found_user_ids):
+                user_ids_to_add = set(expected_user_ids).difference(found_user_ids)
+                for user_id in user_ids_to_add:
+                    conn.execute(
+                        text("insert into galaxy_user(id, active, email, password) values (:id, :active, :email, :pw)"),
+                        [{"id": user_id, "active": True, "email": "e", "pw": "p"}],
+                    )
+
+
+class TestCeleryUserRateLimitIntegration(IntegrationTestCase):
+    """
+    Test feature that limits the number of celery task executions
+    per user per second. This is implemented by the celery_user_rate_limit
+    config parameter whose default value is 0.0 meaning no rate limit.
+    For each of a set of users it submits runs a task num_calls times.
+    For the test to succeed the total duration of completion of all task
+    executions should be equal to the tasks_per_user_per_sec times
+    num_calls-1. This is because tasks for different users should be
+    executed in parallel. In addition, we verify that the total
+    duration for execution of tasks for each user equals the total
+    task duration, again because tasks for different users should
+    run in parallel.
+    """
+
+    def setUp(self):
+        super().setUp()
+
+    def _test_mock_pass_user_id_task(self, users: Iterable[int], num_calls: int, tasks_per_user_per_sec: float):
```

Comment:
This is a rather complicated function that, I think, could benefit from a few comments (it took me some time to understand what it was doing). Maybe add just a few very brief inline comments to each code block, indicating something like `execute tasks for each user, store results / verify results / verify elapsed time / verify elapsed time per user`? I think that would make it much easier to understand.

Reply:
In general, I don't like to clutter the code with inline comments. That is why I created a fairly detailed class level comment. Also, I noticed that the galaxy code base is very sparingly commented in general which has made understanding it very slow going. However, I will add the extra comments because I believe in making code as understandable as possible for others.

Reply:
Thank you! I do believe it'll help. 
(and yes, lots of places in the code base where we should've added more comments, used better variable names, had broken up monster methods, etc.)